_posts/2018-10-11-migrating-from-galen-to-backstop.md

```diff
@@ -0,0 +1,229 @@
+1. Wait for all of the events to resolve.
+1. Wait a bit more (the 100ms time).
+
+Done.
+
+## Switch
+
+For some time we were running both (galen and backstop based) visual regression tests in parallel. After having the backstop one stabilized, we turned the other one off.
+
+## Git repository and storing reference files
+
+One of the things we perceive as crucial is the possibility to clone the repo of our application and with just a few commands run it or execute the tests. It should be self contained.
+
+What about storing reference files inside the repository, then? Should we keep them there? We decided **not to**, and to put them on the server, where all of the tests are executed. Why? For two key reasons.
+
+### 1. Differences in rendering
+
+Despite the fact, that all of the tests are run in the same browser everywhere (headless Chrome), it is not guaranteed that the rendering will look 100% the same on various OSs. E.g. I am using a macOS, my colleague is using some Linux distro. I bet there will be differences. If he were to submit new reference files, tests on my end would fail. So could the ones executed on the pipeline server (yet another OS).
+
+Therefore we decided to keep the files on the pipeline server, where the whole environment is constant.
+
+### 2. Large amount of files
+
+Right now, our reference files weight ~85MB. Let's say something gets changed in the header of the application. Most of the images would have to be refreshed and, say, ~60MB of new ones would be created. What would that mean? Downloading that ~60MB of data on next `git pull`, or whooping ~145MB of data on `git clone`. What if more changes are applied, or reverted? More and more MBs to download. That would be just too much.
+
+Sure, there are ways of bypassing this. One of the solutions is to use [LFS](https://git-lfs.github.com/). Yet, using that extension doesn't fix the no. 1 problem (rendering). We could go a bit further and use [Docker](https://www.docker.com/) to ensure we have the very same environment everywhere, but that just adds too much complexity 'only' for running visual regression tests.
+
+## Final words, learnings, possible improvements
+
+Occasionally we had, and still have, some tests failing from time to time (lazy loaded images, very rarely), but we are satisfied now. The most important thing is that we can ship more often. **We went down from ~30min to less than 5 with our pipeline!**
+
+As we have previously tested our application against various devices, we came to the conclusion, that perhaps we don't to be so precise. Maybe it is OK to have it tested against the most popular browser out there (where most of the modern browsers behave almost identically) and on one device. The time (30min down to 5) and the money saved (from the time to release and dropping of virtual machines from external partners) could be used elsewhere, e.g. to provide more features to our customers. And that is what we did.
```

Comment:
`s/we don't to be/we don't need to be`

Reply:
I would say that main point is that slight visual change (login button unintentionally moved 10 pixels down on less popular browser, but still visible) doesn't cause money loss. Not working `book` button does.